runtime/src/main/java/org/corfudb/runtime/LogReplicationLogicalGroupClient.java

```diff
@@ -0,0 +1,403 @@
+import org.corfudb.runtime.LogReplication.DestinationInfoVal;
+import org.corfudb.runtime.LogReplication.ReplicationModel;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.CorfuStoreEntry;
+import org.corfudb.runtime.collections.Table;
+import org.corfudb.runtime.collections.TableOptions;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.exceptions.TransactionAbortedException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import java.lang.reflect.InvocationTargetException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.corfudb.runtime.view.TableRegistry.CORFU_SYSTEM_NAMESPACE;
+
+/**
+ * A client to interface with log replication utilizing the logical groups replication model.
+ */
+@Slf4j
+public class LogReplicationLogicalGroupClient {
+
+    /**
+     * Key: ClientRegistrationId
+     * Value: ClientRegistrationInfo
+     */
+    public static final String LR_REGISTRATION_TABLE_NAME = "LogReplicationRegistrationTable";
+
+    /**
+     * Key: ClientDestinationInfoKey
+     * Value: DestinationInfoVal
+     */
+    public static final String LR_MODEL_METADATA_TABLE_NAME = "LogReplicationModelMetadataTable";
+
+    private static final ReplicationModel model = ReplicationModel.LOGICAL_GROUPS;
+
+    private final CorfuStore corfuStore;
+    private final Table<ClientRegistrationId, ClientRegistrationInfo, Message> replicationRegistrationTable;
+    private final Table<ClientDestinationInfoKey, DestinationInfoVal, Message> sourceMetadataTable;
+
+    private final ClientRegistrationId clientKey;
+    private final ClientRegistrationInfo clientInfo;
+
+    private final String clientName;
+
+    /**
+     * Constructor for LogReplicationLogicalGroupClient
+     *
+     * @param runtime Corfu Runtime
+     * @param clientName String representation of the client name
+     * @throws InvocationTargetException InvocationTargetException
+     * @throws NoSuchMethodException NoSuchMethodException
+     * @throws IllegalAccessException IllegalAccessException
+     */
+    public LogReplicationLogicalGroupClient(CorfuRuntime runtime, String clientName) throws InvocationTargetException, NoSuchMethodException, IllegalAccessException {
+        Preconditions.checkArgument(isValid(clientName), String.format("%s", "clientName is null or empty."));
+        Table<ClientDestinationInfoKey, DestinationInfoVal, Message> tempSourceMetadataTable;
+        Table<ClientRegistrationId, ClientRegistrationInfo, Message> tempReplicationRegistrationTable;
```

Comment:
why do we need these temp tables? If we do assignment directly, I thought at least it will fall into the catch clause upon exception so we should be fine right?

Reply:
Since the tables are final the compiler throws an error since they may already have been assigned, but shouldn't be a problem making them not final and assigning directly.